### ISSUE TITLE
ipc sendSync - Callback never called & returning event.returnValue inside a callback fails

### ISSUE THREAD

**Issue by RaghuChandrasekaran** [electron/electron#5018]
**OS:**  Windows 10,Windows 7
**Electron Version:**  0.37.4

In my renderer process (web page),I am sending a message synchronously for printing the page as PDF.

In the callback of `window.webContents.printToPDF` I am returning the` returnValue` of event.But the callback is never called.

When attempting the same process asynchronously it works fine.

**Code:**

_Renderer:_
```

var params = {
    "path": undefined,
    "filename": vm.profile._id + " " + vm.profile.firstname
}
let result = ipcRenderer.sendSync('print-profile-pdf', params);
```
_Main Process:_
```

ipcMain.on('print-profile-pdf', function(event, params) {
    printToPdf(params.path, params.filename, event);
});
function printToPdf(path, filename, event) {
    var options = {
        marginsType: 2,
        pageSize: "A4",
        printBackground: false,
        printSelectionOnly: false,
        landscape: false
    };
    let file = (path || app.getPath("desktop")) + "/" + (filename || "profile") + ".pdf";
    previewWindow.webContents.printToPDF(options, function(error, data) {
        if (error) {
            logger.error(error);
            event.returnValue = {
                error: error,
                success: false
            };
            throw error;
        }
        fs.writeFile(file, data, function(error) {
            if (error) {
                logger.error(error);
                event.returnValue = {
                    error: error,
                    success: false
                };
                throw error;
            }
            logger.info("PDF Saved Successfully in " + filename);
            event.returnValue = {
                filename: filename + ".pdf",
                success: true
            };
        });
    });
```
I was able to duplicate the same issue in a fresh setup too.
_**Repo:**_ https://github.com/RaghuChandrasekaran/electron-quick-start

**Reason for sync call:** I don't want to allow the user to intreact until this operation gets successfully executed.

**Comment by MarshallOfSound** [comment]
I personally haven't used used `sendSync` before.  But just as a side opinionated point.

Using synchronous execution of code to block a user from interactions with the app seems like a bad plan.  Especially when printing to PDF it involved, imagine your user has a not so modern CPU and it takes a 5-10 seconds for the PDF to render, your app is completely unresponsive for that time and might even trigger the "Window has stopped responding" messages in Windows.

TLDR:  You should probably use the asynchronous calls and block the user interacting with certain bits of the UI yourself

**Comment by RaghuChandrasekaran** [comment]
@MarshallOfSound I understood what a bad plan it was to use synchronous execution after this issue cropped up :smiley: .Now I have reverted back to the use of async call and used other UI means to restrict user actions.But since this issue (may be expected behaviour) is existing I wanted to have a better insight on why this happening.

**Comment by MarshallOfSound** [comment]
@RaghuChandrasekaran Well as far as I can see, you aren't using `sendSync` correctly.  You are trying to set a return value for a synchronous call in an asynchronous method?  That doesn't seem like it should be or is a thing?

It would be akin to expecting this to work  
```js
const result = (() => {
    this.doSomething((data) => {
        return data;
    });
});
```

Someone might point out a fundamental misunderstanding here, but yeah, that's what I see it as :+1:

**Comment by zcbenz** [comment]
The `printToPDF` API sends messages to the renderer process, while the renderer process is still waiting for the main process to respond the previous message, so this results in a dead lock. I'm afraid this can not be fixed because of the natural of `ipc.sendSync`, I'll update the docs for this.
